### PR TITLE
Add --files-from and --files-from0 options

### PR DESCRIPTION
**Pull request by rotty**
This makes attic usable with custom file listers (e.g. 'find'
invocations) and pre-assembled file lists. Both
newline-delimited (--files-from) and NUL-delimited (--files-from0)
formats are supported.

### CODE REVIEW

**Comment by ThomasWaldmann** [comment]
could you add some unit tests?

**Comment by rotty** [comment]
I'll do so, hopefully in this weekend.

**Comment by rotty** [comment]
I've now added basic unit tests, however, as noted in the commit message, these are not yet very comprehensive. If you decide to merge this branch at this point nevertheless, feel free to squash the two commits (code and unit test).

File: attic/helpers.py [hunk 1]
```diff
@@ -528,6 +528,63 @@ def make_path_safe(path):
...
+    """
+    def __init__(self, mode='r', bufsize=-1, **kwargs):
+        super().__init__(mode=mode, bufsize=bufsize)
+        self._attrs = kwargs
+        self._binary = 'b' in mode
+
+    def __call__(self, string):
+        result = super().__call__(string)
+        # Work around http://bugs.python.org/issue14156
+        if self._binary and result is sys.stdin or result is sys.stdout:
```
Comment (ThomasWaldmann) [inline attic/helpers.py]: is the line above doing what you want?

File: attic/helpers.py [hunk 2]
```diff
@@ -528,6 +528,63 @@ def make_path_safe(path):
...
+            pos = data.find(delim, start)
+            if pos < 0:
+                break
+            yield empty.join(bufs) + data[start:pos] + delim_out
+            start = pos + len(delim)
+            bufs = []
+        if start < len(data):
+            bufs.append(data[start:])
+    if len(bufs) > 0:
+        yield empty.join(bufs)
```
Comment (ThomasWaldmann) [inline attic/helpers.py:566]: please add some unit tests just for this function.
consider edge cases like:
empty input file
file starts/ends with delim
2 delims directly following each other

File: attic/helpers.py (hunk 1 above)
Comment (rotty) [inline attic/helpers.py]: Manual tests using stdin work (this is my primary use case, after all), however, to add automatic tests, one would need to extend 'ArchiverTestCaseBase.attic()' to emulate sys.stdin.buffer to enable binary input from stdin.

Comment (ThomasWaldmann) [inline attic/helpers.py]: do you mean it like this:  if (self._binary and result is sys.stdin) or result is sys.stdout: ?

Comment (rotty) [inline attic/helpers.py]: TW notifications@github.com writes:

> > +class FileType(argparse.FileType):
> > +
> > -    [...]
> >   +
> > -    def **call**(self, string):
> > -        result = super().**call**(string)
> > -        # Work around http://bugs.python.org/issue14156
> > -        if self._binary and result is sys.stdin or result is sys.stdout:
> 
> do you mean it like this:  if (self._binary and result is sys.stdin)
> or result is sys.stdout: ?
> 
> Oh, good catch, there is a set of parens missing (should be 'if
> self._binary and (result is sys.stdin or result is
> sys.stdout)'. Obviously, I didn't test with sys.stdout. I'll fix this,
> along with adding a unit test for this class.

Thanks, Rotty

Comment (rotty) [inline attic/helpers.py]: Another thing that occurred to me, and which i'm not exactly sure what to do about, is that the code in `FileType.__call__()` monkey-patches (i.e., adds attributes to) `sys.stdin` or `sys.stderr`, if they are the result of `super().__call__()`.

I think this monkey-patching is fine for newly-created file objects, but less so for `sys.stdin` and `sys.stdout`. It might however be acceptable in the context of an application (as opposed to a general-purpose library). Do you have an opinion on that topic?

**Comment by rotty** [comment]
Oh, that Travis CI failure is sneaky: I can reproduce it locally, but it only happens when running `python3 -m attic.testsuite.run -b`, but not without `-b` ("Buffer stdout and stderr during tests").
